File: Dopamine.SettingsModule/ViewModels/SettingsBehaviourViewModel.cs
using Digimezzo.Utilities.Helpers;
using Digimezzo.Utilities.Settings;
using Digimezzo.Utilities.Utils;
using Dopamine.Common.Prism;
using Prism.Events;
using Prism.Mvvm;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows;

namespace Dopamine.SettingsModule.ViewModels
{
    public class SettingsBehaviourViewModel : BindableBase
    {
        #region Variables
        private IEventAggregator eventAggregator;
        private bool checkBoxShowTrayIconChecked;
        private bool checkBoxMinimizeToTrayChecked;
        private bool checkBoxFollowTrackChecked;
        private bool checkBoxCloseToTrayChecked;
        private bool checkBoxEnableRatingChecked;
        private bool checkBoxEnableLoveChecked;
        private bool checkBoxSaveRatingInAudioFilesChecked;
        private bool checkBoxShowRemoveFromDiskChecked;
        private ObservableCollection<NameValue> scrollVolumePercentages;
        private NameValue selectedScrollVolumePercentage;
        #endregion

        #region Properties
        public bool CheckBoxShowTrayIconChecked
        {
            get { return this.checkBoxShowTrayIconChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowTrayIcon", value);
                SetProperty<bool>(ref this.checkBoxShowTrayIconChecked, value);
                Application.Current.Dispatcher.Invoke(() => this.eventAggregator.GetEvent<SettingShowTrayIconChanged>().Publish(value));
            }
        }

        public bool CheckBoxMinimizeToTrayChecked
        {
            get { return this.checkBoxMinimizeToTrayChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "MinimizeToTray", value);
                SetProperty<bool>(ref this.checkBoxMinimizeToTrayChecked, value);
            }
        }

        public bool CheckBoxCloseToTrayChecked
        {
            get { return this.checkBoxCloseToTrayChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "CloseToTray", value);
                SetProperty<bool>(ref this.checkBoxCloseToTrayChecked, value);
            }
        }

        public bool CheckBoxFollowTrackChecked
        {
            get { return this.checkBoxFollowTrackChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "FollowTrack", value);
                SetProperty<bool>(ref this.checkBoxFollowTrackChecked, value);
            }
        }

        public bool CheckBoxEnableRatingChecked
        {
            get { return this.checkBoxEnableRatingChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "EnableRating", value);
                SetProperty<bool>(ref this.checkBoxEnableRatingChecked, value);
                Application.Current.Dispatcher.Invoke(() => this.eventAggregator.GetEvent<SettingEnableRatingChanged>().Publish(value));
            }
        }

        public bool CheckBoxEnableLoveChecked
        {
            get { return this.checkBoxEnableLoveChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "EnableLove", value);
                SetProperty<bool>(ref this.checkBoxEnableLoveChecked, value);
                Application.Current.Dispatcher.Invoke(() => this.eventAggregator.GetEvent<SettingEnableLoveChanged>().Publish(value));
            }
        }

        public bool CheckBoxSaveRatingInAudioFilesChecked
        {
            get { return this.checkBoxSaveRatingInAudioFilesChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "SaveRatingToAudioFiles", value);
                SetProperty<bool>(ref this.checkBoxSaveRatingInAudioFilesChecked, value);
            }
        }

        public bool CheckBoxShowRemoveFromDiskChecked
        {
            get { return this.checkBoxShowRemoveFromDiskChecked; }
            set
            {
                SettingsClient.Set<bool>("Behaviour", "ShowRemoveFromDisk", value);
                SetProperty<bool>(ref this.checkBoxShowRemoveFromDiskChecked, value);
                Application.Current.Dispatcher.Invoke(() => this.eventAggregator.GetEvent<SettingShowRemoveFromDiskChanged>().Publish(value));
            }
        }

    public ObservableCollection<NameValue> ScrollVolumePercentages
        {
            get { return this.scrollVolumePercentages; }
            set { SetProperty<ObservableCollection<NameValue>>(ref this.scrollVolumePercentages, value); }
        }

        public NameValue SelectedScrollVolumePercentage
        {
            get { return this.selectedScrollVolumePercentage; }
            set
            {
                SettingsClient.Set<int>("Behaviour", "ScrollVolumePercentage", value.Value);
                SetProperty<NameValue>(ref this.selectedScrollVolumePercentage, value);
            }
        }
        #endregion

        #region Construction
        public SettingsBehaviourViewModel(IEventAggregator eventAggregator)
        {
            this.eventAggregator = eventAggregator;

            this.GetCheckBoxesAsync();
            this.GetScrollVolumePercentagesAsync();
        }
        #endregion

        #region Private
        private async void GetCheckBoxesAsync()
        {
            await Task.Run(() =>
            {
                this.CheckBoxShowTrayIconChecked = SettingsClient.Get<bool>("Behaviour", "ShowTrayIcon");
                this.CheckBoxMinimizeToTrayChecked = SettingsClient.Get<bool>("Behaviour", "MinimizeToTray");
                this.CheckBoxCloseToTrayChecked = SettingsClient.Get<bool>("Behaviour", "CloseToTray");
                this.CheckBoxFollowTrackChecked = SettingsClient.Get<bool>("Behaviour", "FollowTrack");
                this.CheckBoxEnableRatingChecked = SettingsClient.Get<bool>("Behaviour", "EnableRating");
                this.CheckBoxEnableLoveChecked = SettingsClient.Get<bool>("Behaviour", "EnableLove");
                this.CheckBoxShowRemoveFromDiskChecked = SettingsClient.Get<bool>("Behaviour", "ShowRemoveFromDisk");
                this.CheckBoxSaveRatingInAudioFilesChecked = SettingsClient.Get<bool>("Behaviour", "SaveRatingToAudioFiles");
            });
        }

        private async void GetScrollVolumePercentagesAsync()
        {
            var localScrollVolumePercentages = new ObservableCollection<NameValue>();

            await Task.Run(() =>
            {
                localScrollVolumePercentages.Add(new NameValue { Name = "1 %", Value = 1 });
                localScrollVolumePercentages.Add(new NameValue { Name = "2 %", Value = 2 });
                localScrollVolumePercentages.Add(new NameValue { Name = "5 %", Value = 5 });
                localScrollVolumePercentages.Add(new NameValue { Name = "10 %", Value = 10 });
                localScrollVolumePercentages.Add(new NameValue { Name = "15 %", Value = 15 });
                localScrollVolumePercentages.Add(new NameValue { Name = "20 %", Value = 20 });
            });

            this.ScrollVolumePercentages = localScrollVolumePercentages;

            NameValue localSelectedScrollVolumePercentage = null;
            await Task.Run(() => localSelectedScrollVolumePercentage = this.ScrollVolumePercentages.Where((svp) => svp.Value == SettingsClient.Get<int>("Behaviour", "ScrollVolumePercentage")).Select((svp) => svp).First());
            this.SelectedScrollVolumePercentage = localSelectedScrollVolumePercentage;
        }
        #endregion
    }
}

File: Dopamine.SettingsModule/ViewModels/SettingsBehaviourViewModel.vb
Imports Microsoft.Practices.Prism.Mvvm
Imports Dopamine.Core.Settings
Imports System.Collections.ObjectModel
Imports Dopamine.Core.Utils
Imports Dopamine.Core.Base
Imports Microsoft.Practices.Prism.PubSubEvents
Imports Dopamine.Core.Prism

Namespace ViewModels
    Public Class SettingsBehaviourViewModel
        Inherits BindableBase

#Region "Variables"
        Private mEventAggregator As IEventAggregator
        Private mCheckBoxShowTrayIconChecked As Boolean
        Private mCheckBoxMinimizeToTrayChecked As Boolean
        Private mCheckBoxFollowTrackChecked As Boolean
        Private mCheckBoxCloseToTrayChecked As Boolean
        Private mCheckBoxEnableRatingChecked As Boolean
        Private mCheckBoxUseStarRatingChecked As Boolean
        Private mCheckBoxSaveRatingInAudioFilesChecked As Boolean
#End Region

#Region "Properties"
        Public Property CheckBoxShowTrayIconChecked() As Boolean
            Get
                Return mCheckBoxShowTrayIconChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "ShowTrayIcon", value)
                SetProperty(Of Boolean)(mCheckBoxShowTrayIconChecked, value)
                Application.Current.Dispatcher.Invoke(Sub() mEventAggregator.GetEvent(Of SettingShowTrayIconChanged).Publish(value))
            End Set
        End Property

        Public Property CheckBoxMinimizeToTrayChecked() As Boolean
            Get
                Return mCheckBoxMinimizeToTrayChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "MinimizeToTray", value)
                SetProperty(Of Boolean)(mCheckBoxMinimizeToTrayChecked, value)
            End Set
        End Property

        Public Property CheckBoxCloseToTrayChecked() As Boolean
            Get
                Return mCheckBoxCloseToTrayChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "CloseToTray", value)
                SetProperty(Of Boolean)(mCheckBoxCloseToTrayChecked, value)
            End Set
        End Property

        Public Property CheckBoxFollowTrackChecked() As Boolean
            Get
                Return mCheckBoxFollowTrackChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "FollowTrack", value)
                SetProperty(Of Boolean)(mCheckBoxFollowTrackChecked, value)
            End Set
        End Property

        Public Property CheckBoxEnableRatingChecked() As Boolean
            Get
                Return mCheckBoxEnableRatingChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "EnableRating", value)
                SetProperty(Of Boolean)(mCheckBoxEnableRatingChecked, value)
                Application.Current.Dispatcher.Invoke(Sub() mEventAggregator.GetEvent(Of SettingEnableRatingChanged)().Publish(value))
            End Set
        End Property

        Public Property CheckBoxUseStarRatingChecked() As Boolean
            Get
                Return mCheckBoxUseStarRatingChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "UseStarRating", value)
                SetProperty(Of Boolean)(mCheckBoxUseStarRatingChecked, value)
                Application.Current.Dispatcher.Invoke(Sub() mEventAggregator.GetEvent(Of SettingUseStarRatingChanged)().Publish(value))
            End Set
        End Property

        Public Property CheckBoxSaveRatingInAudioFilesChecked() As Boolean
            Get
                Return mCheckBoxSaveRatingInAudioFilesChecked
            End Get
            Set(ByVal value As Boolean)
                XmlSettingsClient.Instance.Set(Of Boolean)("Behaviour", "SaveRatingToAudioFiles", value)
                SetProperty(Of Boolean)(mCheckBoxSaveRatingInAudioFilesChecked, value)
            End Set
        End Property
#End Region

#Region "Construction"
        Public Sub New(iEventAggregator As IEventAggregator)

            mEventAggregator = IEventAggregator

            Me.GetCheckBoxesAsync()
        End Sub
#End Region

#Region "Private"
        Private Async Sub GetCheckBoxesAsync()

            Await Task.Run(Sub()
                               Me.CheckBoxShowTrayIconChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "ShowTrayIcon")
                               Me.CheckBoxMinimizeToTrayChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "MinimizeToTray")
                               Me.CheckBoxCloseToTrayChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "CloseToTray")
                               Me.CheckBoxFollowTrackChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "FollowTrack")
                               Me.CheckBoxEnableRatingChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "EnableRating")
                               Me.CheckBoxUseStarRatingChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "UseStarRating")
                               Me.CheckBoxSaveRatingInAudioFilesChecked = XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "SaveRatingToAudioFiles")
                           End Sub)
        End Sub
#End Region
    End Class
End Namespace